Fixes for Example Architecture

jnfoster opened the pull request:
Addresses mistakes noted in #340

jnfoster commented:
@tomrodeheffer can you take a look also?

cc10512 commented on p4-16/spec/P4-16-spec.mdk:
```diff
@@ -5585,48 +5585,43 @@ control EPipe<T, IH, OH>(in IH inputHeaders,
+  instantiated with the same type in `Switch`. In contrast, note that
+  the `Ingress` type `IH` and the `Egress` type `IH` may be
+  different. To force them to be the same type, we could instead
+  declare `IH` and `OH` at the switch level: `package Switch<IH, OH, T>`.
```
you also have to change the argument signatures from `Ingress<T,_,_> ingress, Egress<T,_,_> egress`, to `Ingress<T,IH,OH> ingress, Ingress<T,IH,OH> egress`, as it is not clear how the defaults will propagate to arguments.

jnfoster replied:
The compiler will fill these in when Switch itself is instantiated.

jnfoster replied:
Just to confirm, the compiler accepts this program...
```
#include <core.p4>
    
typedef bit<4> PortId;
    
struct InControl {
    PortId inputPort;
}
    
struct OutControl {
    PortId outputPort;
}    
        
parser Parser<IH>(packet_in b, out IH parsedHeaders);
    
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
    
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);
control Deparser<OH>(in OH outputHeaders, packet_out b);
    
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<T, IH, OH> map,
                           Deparser<OH> d);
    
package Egress<T, IH, OH>(Parser<IH> p,
                          EPipe<T, IH, OH> map,
                          Deparser<OH> d);
    
package Switch<T>(Ingress<T, _, _> ingress, Egress<T, _, _> egress);
```
If I wanted to make all of the type variables explicit, I'd have to parameterize `Switch` on additional type variables:
```
#include <core.p4>
    
typedef bit<4> PortId;
    
struct InControl {
    PortId inputPort;
}
    
struct OutControl {
    PortId outputPort;
}    
        
parser Parser<IH>(packet_in b, out IH parsedHeaders);
    
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
    
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);
    
control Deparser<OH>(in OH outputHeaders, packet_out b);
    
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<T, IH, OH> map,
                           Deparser<OH> d);
    
package Egress<T, IH, OH>(Parser<IH> p,
                          EPipe<T, IH, OH> map,
                          Deparser<OH> d);
    
package Switch<T,H1,H2,H3,H4>(Ingress<T, H1, H2> ingress, Egress<T, H3, H4> egress);
```
Note that filling in `IH` and `OH` for the `_`s doesn't compile -- first, those type variables are unbound, and second, that would equate two types that are intended to be independent (per the comments above).

cc10512 replied:
so then where does this leave your statement that you can pull the IH and OH into the package decl: ```package Switch<IH, OH, T>```?

jnfoster replied:
Sorry, the second example above was to preserve the same (independent) design. 

To force them to be the same, we could write this:
```
#include <core.p4>
    
typedef bit<4> PortId;
    
struct InControl {
    PortId inputPort;
}
    
struct OutControl {
    PortId outputPort;
}    
        
parser Parser<IH>(packet_in b, out IH parsedHeaders);
    
control IPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         out OH outputHeaders,
                         out T toEgress,
                         out OutControl outCtrl);
    
control EPipe<T, IH, OH>(in IH inputHeaders,
                         in InControl inCtrl,
                         in T fromIngress,
                         out OH outputHeaders,
                         out OutControl outCtrl);
    
control Deparser<OH>(in OH outputHeaders, packet_out b);
    
package Ingress<T, IH, OH>(Parser<IH> p,
                           IPipe<T, IH, OH> map,
                           Deparser<OH> d);
    
package Egress<T, IH, OH>(Parser<IH> p,
                          EPipe<T, IH, OH> map,
                          Deparser<OH> d);
    
package Switch<T,IH,OH>(Ingress<T, IH,OH> ingress, Egress<T, IH,OH> egress);
```

cc10512 replied:
Yes, that was what I thought and tried to express in the initial comment :)

jnfoster replied:
Now I'm getting confused. I think both are legal programs but they mean different things... Let's just sort it out in person tomorrow?

jnfoster replied:
Done.

mbudiu-vmw commented:
Please add these programs as tests for the compiler. I expected that the one in the text already exists.

jnfoster commented:
@cc10512 I added the complete statement for the example, as requested.

@mbudiu-vmw I added tests for p4c https://github.com/p4lang/p4c/pull/777

I think we're good to merge, yes?
